File: store/src/main/java/com/nytimes/android/external/store/base/impl/RealStoreBuilder.java
package com.nytimes.android.external.store.base.impl;


import com.nytimes.android.external.cache.Cache;
import com.nytimes.android.external.store.base.DiskRead;
import com.nytimes.android.external.store.base.DiskWrite;
import com.nytimes.android.external.store.base.Fetcher;
import com.nytimes.android.external.store.base.Parser;
import com.nytimes.android.external.store.base.Persister;
import com.nytimes.android.external.store.util.KeyParser;
import com.nytimes.android.external.store.util.NoKeyParser;
import com.nytimes.android.external.store.util.NoopParserFunc;
import com.nytimes.android.external.store.util.NoopPersister;

import java.util.ArrayList;
import java.util.List;

import javax.annotation.Nonnull;

import rx.Observable;

/**
 * Builder where there parser is used.
 */
public class RealStoreBuilder<Raw, Parsed, Key> {
    private final List<KeyParser> parsers = new ArrayList<>();
    private Persister<Raw, Key> persister;
    private Cache<Key, Observable<Parsed>> memCache;
    private Fetcher<Raw, Key> fetcher;

    @SuppressWarnings("PMD.UnusedPrivateField") //remove when it is implemented...
    private StalePolicy stalePolicy = StalePolicy.UNSPECIFIED;

    @Nonnull
    public static <Raw, Parsed, Key> RealStoreBuilder<Raw, Parsed, Key> builder() {
        return new RealStoreBuilder<>();
    }

    @Nonnull
    public RealStoreBuilder<Raw, Parsed, Key> fetcher(final @Nonnull Fetcher<Raw, Key> fetcher) {
        this.fetcher = fetcher;
        return this;
    }

    @Nonnull
    public RealStoreBuilder<Raw, Parsed, Key> persister(final @Nonnull Persister<Raw, Key> persister) {
        this.persister = persister;
        return this;
    }

    @Nonnull
    public RealStoreBuilder<Raw, Parsed, Key> persister(final @Nonnull DiskRead<Raw, Key> diskRead,
                                                        final @Nonnull DiskWrite<Raw, Key> diskWrite) {
        persister = new Persister<Raw, Key>() {
            @Nonnull
            @Override
            public Observable<Raw> read(@Nonnull Key key) {
                return diskRead.read(key);
            }

            @Nonnull
            @Override
            public Observable<Boolean> write(@Nonnull Key key, @Nonnull Raw raw) {
                return diskWrite.write(key, raw);
            }
        };
        return this;
    }

    @Nonnull
    public RealStoreBuilder<Raw, Parsed, Key> parser(final @Nonnull Parser<Raw, Parsed> parser) {
        this.parsers.clear();
        this.parsers.add(new NoKeyParser<>(parser));
        return this;
    }

    @Nonnull
    public RealStoreBuilder<Raw, Parsed, Key> parser(final @Nonnull KeyParser<Key, Raw, Parsed> parser) {
        this.parsers.clear();
        this.parsers.add(parser);
        return this;
    }


    @Nonnull
    @SuppressWarnings("PMD.AvoidInstantiatingObjectsInLoops")
    public RealStoreBuilder<Raw, Parsed, Key> parsers(final @Nonnull List<Parser> parsers) {
        this.parsers.clear();
        for (Parser parser : parsers) {
            this.parsers.add(new NoKeyParser<>(parser));
        }
        return this;
    }

    @Nonnull
    public RealStoreBuilder<Raw, Parsed, Key> memory(Cache<Key, Observable<Parsed>> memCache) {
        this.memCache = memCache;
        return this;
    }

    //Store will backfill the disk cache anytime a record is stale
    //User will still get the stale record returned to them
    public RealStoreBuilder<Raw, Parsed, Key> refreshOnStale() {
        stalePolicy = StalePolicy.REFRESH_ON_STALE;
        return this;
    }

    //Store will try to get network source when disk data is stale
    //if network source throws error or is empty, stale disk data will be returned
    @Nonnull
    public RealStoreBuilder<Raw, Parsed, Key> networkBeforeStale() {
        stalePolicy = StalePolicy.NETWORK_BEFORE_STALE;
        return this;
    }

    @Nonnull
    public Store<Parsed, Key> open() {
        if (persister == null) {
            persister = new NoopPersister<>();
        }

        if (parsers.isEmpty()) {
            parser(new NoopParserFunc<Raw, Parsed>());
        }

        KeyParser<Key, Raw, Parsed> multiParser = new MultiParser<>(parsers);
        RealInternalStore<Raw, Parsed, Key> realInternalStore;

        if (memCache == null) {
            realInternalStore = new RealInternalStore<>(fetcher, persister, multiParser, stalePolicy);
        } else {
            realInternalStore = new RealInternalStore<>(fetcher, persister, multiParser, memCache, stalePolicy);
        }

        return new RealStore<>(realInternalStore);
    }
}

File: store/src/main/java/com/nytimes/android/external/store3/base/impl/RealStoreBuilder.kt
package com.nytimes.android.external.store3.base.impl


import com.nytimes.android.external.store3.base.*
import com.nytimes.android.external.store3.util.KeyParser
import com.nytimes.android.external.store3.util.NoKeyParser
import com.nytimes.android.external.store3.util.NoopParserFunc
import com.nytimes.android.external.store3.util.NoopPersister
import java.util.*


/**
 * Builder where there parser is used.
 */
class RealStoreBuilder<Raw, Parsed, Key> {
    private val parsers = ArrayList<KeyParser<Any?, Any?, Any?>>()
    private var persister: Persister<Raw, Key>? = null
    private var fetcher: Fetcher<Raw, Key>? = null
    private var memoryPolicy: MemoryPolicy? = null

    private//remove when it is implemented...
    var stalePolicy = StalePolicy.UNSPECIFIED

    fun fetcher(fetcher: Fetcher<Raw, Key>): RealStoreBuilder<Raw, Parsed, Key> = apply {
        this.fetcher = fetcher
    }

    fun fetcher(fetcher: suspend (Key) -> Raw): RealStoreBuilder<Raw, Parsed, Key> = apply {
        this.fetcher = object : Fetcher<Raw, Key> {
            override suspend fun fetch(key: Key): Raw {
                return fetcher(key)
            }
        }
    }

    fun persister(persister: Persister<Raw, Key>): RealStoreBuilder<Raw, Parsed, Key> = apply {
        this.persister = persister
    }

    fun persister(diskRead: DiskRead<Raw, Key>,
                  diskWrite: DiskWrite<Raw, Key>): RealStoreBuilder<Raw, Parsed, Key> = apply {
        persister = object : Persister<Raw, Key> {
            override suspend fun read(key: Key): Raw? =
                    diskRead.read(key)

            override suspend fun write(key: Key, raw: Raw): Boolean =
                    diskWrite.write(key, raw)
        }
    }

    fun parser(parser: Parser<Raw, Parsed>): RealStoreBuilder<Raw, Parsed, Key> = apply {
        this.parsers.clear()
        this.parsers.add(NoKeyParser(parser as Parser<Any?, Any?>))
    }

    fun parser(parser: suspend (Raw) -> Parsed): RealStoreBuilder<Raw, Parsed, Key> =
            parser(NoKeyParser(object : Parser<Raw, Parsed> {
                override suspend fun apply(raw: Raw): Parsed {
                    return parser(raw)
                }
            }))

    fun parser(parser: KeyParser<Key, Raw, Parsed>): RealStoreBuilder<Raw, Parsed, Key> = apply {
        this.parsers.clear()
        this.parsers.add(parser as KeyParser<Any?, Any?, Any?>)
    }

    fun parsers(parsers: List<Parser<*, *>>): RealStoreBuilder<Raw, Parsed, Key> = apply {
        this.parsers.clear()
        this.parsers.addAll(parsers.map { NoKeyParser<Any?, Any?, Any?>(it as Parser<Any?, Any?>) })
    }

    fun memoryPolicy(memoryPolicy: MemoryPolicy): RealStoreBuilder<Raw, Parsed, Key> = apply {
        this.memoryPolicy = memoryPolicy
    }

    //Store will backfill the disk cache anytime a record is stale
    //User will still get the stale record returned to them
    fun refreshOnStale(): RealStoreBuilder<Raw, Parsed, Key> = apply {
        stalePolicy = StalePolicy.REFRESH_ON_STALE
    }

    //Store will try to get network source when disk data is stale
    //if network source throws error or is empty, stale disk data will be returned
    fun networkBeforeStale(): RealStoreBuilder<Raw, Parsed, Key> = apply {
        stalePolicy = StalePolicy.NETWORK_BEFORE_STALE
    }

    fun open(): Store<Parsed, Key> {
        if (persister == null) {
            persister = NoopPersister.create(memoryPolicy)
        }

        if (parsers.isEmpty()) {
            parser(NoopParserFunc())
        }

        val multiParser = MultiParser<Key, Raw, Parsed>(parsers)

        val realInternalStore: RealInternalStore<Raw, Parsed, Key> = RealInternalStore(fetcher!!, persister!!, multiParser, memoryPolicy, stalePolicy)

        return RealStore<Parsed, Key>(realInternalStore)
    }

    companion object {

        fun <Raw, Parsed, Key> builder(): RealStoreBuilder<Raw, Parsed, Key> {
            return RealStoreBuilder()
        }
    }
}
